scrapy/utils/misc.py

```diff
@@ -162,3 +165,40 @@ def set_environ(**kwargs):
                 del os.environ[k]
             else:
                 os.environ[k] = v
+
+
+if not six.PY2:
+    from functools import lru_cache
+    from textwrap import dedent
+
+    @lru_cache(maxsize=128)
```

Comment:
My only concern is this cache function: I wonder how it works with e.g. partial functions passed as callbacks - sometimes people use them to pass additional data. Won't it keep bound arguments of these functions for a long time? E.g. imagine something like `Request(url, partial(self.parse2, prev_response=response))` - would ~128 responses be kept alive in is_generator_with_return_value cache after requests are processed? I haven't check it, but I'm not sure lru_key uses weakrefs for keys by default.

Writing such code in spiders is not a good style, but here the tradeoff is a nice-to-have warning vs increased memory usage (and probably a failure) of some spiders.

Reply:
Indeed, `lru_cache` does keep references to the cached objects (https://bugs.python.org/issue19859).
I'm working on a weak referenced cache implementation (based on `scrapy.utils.datatypes.LocalCache`) that does not add much overhead IMHO, will update shortly.

Reply:
Updated, please see [`scrapy.utils.datatypes.LocalWeakReferencedCache`](https://github.com/scrapy/scrapy/pull/3869/files#diff-2f75773d25e485f98def0ba43b0ce78eR324)